### DISCUSSION TITLE
Failed to connect secret provisioning with Intel SGX and Gramine-PyTorch

### DISCUSSION THREAD

**Discussion by mrinal054** [gramineproject/gramine#1629] [Q&A]
Hello,

I was doing PyTorch privacy-preserving machine learning applications with Intel SGX and Gramine, as described in [Gramine’s GitHub](https://github.com/gramineproject/gramine/blob/master/Documentation/tutorials/pytorch/index.rst).

I was able to execute the deep learning model inside the enclave without remote attestation.

However, I am facing problems establishing the remote attestation. Here is the error I am receiving when I am trying to connect to the DCAP server - 

```python
~/Gramine_project/examples/pytorch-confidential$ gramine-sgx ./pytorch pytorchexample.py
Gramine is starting. Parsing TOML manifest file, this may take some time...
-----------------------------------------------------------------------------------------------------------------------
Gramine detected the following insecure configurations:

  - loader.insecure__use_cmdline_argv = true   (forwarding command-line args from untrusted host to the app)

Gramine will continue application execution, but this configuration must not be used in production!
-----------------------------------------------------------------------------------------------------------------------

secret_provision_start: Secret Provisioning could not connect to any of the servers specified in SECRET_PROVISION_SERVERS; last mbedTLS error was -68
secret_provision_common_close: Secret Provisioning failed during connection close with mbedTLS error -28928
WARNING: Closing the secret-prov context failed with error -1.
secret_provision_constructor: Secret provisioning failed, terminating the whole process

```

Here is my server side - 

![dcap_server](https://github.com/gramineproject/gramine/assets/44781227/ff0ac9a5-cd19-4417-819a-c926230816f2)

Here is my user side - 

![client_side](https://github.com/gramineproject/gramine/assets/44781227/4d4f8176-b617-4195-8084-7006b1926893)

I have changed the manifest file as instructed by Gramine. Here is the changes that I made to established the connect and DCAP - 

```python
sgx.trusted_files = [
  "file:{{ entrypoint }}",
  "file:{{ gramine.libos }}",
  "file:{{ gramine.runtimedir() }}/",
  "file:/usr/lib/",
  "file:{{ arch_libdir }}/",
  "file:/usr/{{ arch_libdir }}/",
{% for path in python.get_sys_path(entrypoint) %}
  "file:{{ path }}{{ '/' if path.is_dir() else '' }}",
{% endfor %}

  "file:pytorchexample.py",
  "file:ssl/ca.crt",
]

sys.enable_extra_runtime_domain_names_conf = true
sgx.remote_attestation = "dcap"
loader.env.LD_PRELOAD = "libsecret_prov_attest.so"
loader.env.SECRET_PROVISION_CONSTRUCTOR = "1"
loader.env.SECRET_PROVISION_SET_KEY = "default"
loader.env.SECRET_PROVISION_CA_CHAIN_PATH = "ssl/ca.crt"
loader.env.SECRET_PROVISION_SERVERS = "localhost:4433"
```

I am using Gramine 1.5. Here is the version details - 

```python
Package: gramine
Version: 1.5
Priority: optional
Section: misc
Maintainer: Wojtek Porczyk <woju@invisiblethingslab.com>
Installed-Size: 21.9 MB
Depends: libcurl4 (>= 7.58), libprotobuf-c1, python3, python3-click, python3-cryptography, python3-jinja2, python3-pyelftools, python3-tomli (>= 1.1.0), python3-tomli-w (>= 0.4.0)
Recommends: gramine-ratls-dcap, gramine-ratls-epid
Conflicts: gramine-oot
Breaks: gramine-dcap (<< 1.4~)
Replaces: gramine-dcap (<< 1.4~)
Homepage: https://gramine.readthedocs.io/
Download-Size: 3,351 kB
APT-Manual-Installed: yes
APT-Sources: https://packages.gramineproject.io jammy/main amd64 Packages
Description: A lightweight usermode guest OS designed to run a single Linux application
```

Please note that I have also tried with setting the RA_TLS_ALLOW_OUTDATED_TCB_INSECURE to 1.

As the complete Gramine log is quite long, I am attaching a text file that contains the entire log for loader.log_level = "all".

[log_detail.txt](https://github.com/gramineproject/gramine/files/13184092/log_detail.txt)

Could you please suggest how to resolve the issue?

**Comment by dimakuv** [comment]
Hm, a bit strange.

Can you add the following item to the `sgx.trusted_files` list:
```
sgx.trusted_files = [
   ...
   "file:/etc/hosts",
]
```

and then rebuild the PyTorch example: `make clean; make SGX=1` and try again?

**Reply by dimakuv** [reply]
Sorry, forgot to mention. You'll also need to extend `fs.mounts` list:
```
fs.mounts = [
   ...
   { path = "/etc", uri = "file:/etc" },
]
```

**Reply by mrinal054** [reply]
Thank you for your reply, and sorry for my late response.

After adding those items, it seems like it can recognize the localhost now. However, still it can't establish the secret provisioning. 

Here is what I got from the server side:
```python
~/Gramine_project/examples/pytorch-confidential/secret_prov_pf$ --- Reading the master key for encrypted files from '../files/wrap_key' ---
--- Starting the Secret Provisioning server on port 4433 ---
ra_tls_verify_callback: sgx_qv_verify_quote failed: 57372
client_connection: Secret Provisioning failed during mbedtls_ssl_handshake with error -12288
client_connection: ra_tls_verify_callback_results:
    attestation_scheme=2, err_loc=3, 
client_connection:     dcap.func_verify_quote_result=0xe01c, dcap.quote_verification_result=0xa006
```

And, here is what I got from the user end:
```python
~/Gramine_project/examples/pytorch-confidential$ gramine-sgx ./pytorch pytorchexample.py
Gramine is starting. Parsing TOML manifest file, this may take some time...
-----------------------------------------------------------------------------------------------------------------------
Gramine detected the following insecure configurations:

  - loader.insecure__use_cmdline_argv = true   (forwarding command-line args from untrusted host to the app)

Gramine will continue application execution, but this configuration must not be used in production!
-----------------------------------------------------------------------------------------------------------------------

secret_provision_start: Secret Provisioning failed during mbedtls_ssl_handshake with error -80
secret_provision_constructor: Secret provisioning failed, terminating the whole process
```
I have also attached the detailed log to this reply. 
[log_detail2.txt](https://github.com/gramineproject/gramine/files/13246220/log_detail2.txt)

Thank you in advance.

**Reply by kailun-qin** [reply]
@mrinal054, `0xe01c` indicated a DCAP `SGX_QL_QUOTE_CERTIFICATION_DATA_UNSUPPORTED` error.

Have you installed and configured the DCAP infra (specifically QPL and PCCS) correctly in your env (by following e.g., https://www.intel.com/content/www/us/en/developer/articles/guide/intel-software-guard-extensions-data-center-attestation-primitives-quick-install-guide.html)?

**Reply by taserghar** [reply]
Hi @kailun-qin , me and @mrinal054 are working together. I am encountering an error while trying to install PCCS. The error message reads as follows:

> Setting up npm (8.5.1~ds-1) ...
> Setting up sgx-dcap-pccs (1.19.100.3-jammy1) ...
> Checking nodejs version ...
> **The minimum node.js version required is v16.20.0. Installation aborted.**  
> dpkg: error processing package sgx-dcap-pccs (--configure):
>  installed sgx-dcap-pccs package post-installation script subprocess returned error exit status 1
> Processing triggers for libc-bin (2.35-0ubuntu3.4) ...
> Processing triggers for man-db (2.10.2-1) ...
> Errors were encountered while processing:
>  sgx-dcap-pccs
> E: Sub-process /usr/bin/dpkg returned an error code (1)

If I execute "sudo apt-get -y install nodejs", it installs a lower version. However, PCCS requires v16.20.0, so I remove nodejs and manually installed it by downloading the binary from here.
>https://nodejs.org/en/blog/release/v16.20.0

Here is the log:

> root@SGX-server:/home/dhar/downloads# cp -r node-v16.20.0-linux-x64/{bin,include,lib,share} /usr/
> root@SGX-server:/home/dhar/downloads# export PATH=/usr/node-v16.20.0-linux-x64/bin:$PATH
> root@SGX-server:/home/dhar/downloads# node --version
> 
> v16.20.0

After attempting to reinstall the PCCS with "apt install sgx-dcap-pccs", I encountered the same error as before. Additionally, the node version was automatically changed to v12.

> ....
> **The minimum node.js version required is v16.20.0. Installation aborted.**  
> dpkg: error processing package sgx-dcap-pccs (--configure):
>  installed sgx-dcap-pccs package post-installation script subprocess returned error exit status 1
> Processing triggers for libc-bin (2.35-0ubuntu3.4) ...
> Processing triggers for man-db (2.10.2-1) ...
> Errors were encountered while processing:
>  sgx-dcap-pccs
> E: Sub-process /usr/bin/dpkg returned an error code (1)
> root@SGX-server:/home/dhar/downloads# node -v
> v12.22.9

Could you please shed some light on why the node version is being automatically downgraded during the PCCS installation script? 

Thank you.
Mubashwir

**Reply by dimakuv** [reply]
@taserghar Are you using an old Ubuntu?

Anyway, I think that manual download & installation of Node.js is a bad idea.

I suggest to use `snap` installation instead: https://snapcraft.io/node. You'll need to install the snap package manager first, and then install via `sudo snap install node`. This should install the latest version, without a need for you to do anything additional.

Just make sure to remove the previous installations of Node.js before this.

**Reply by taserghar** [reply]
Hi @dimakuv, Thanks for the reply. I am using Ubuntu 22.04.
I installed the node using Snap, and it installed version 20.x But then PCCS still shows the same error. Here are the detail logs:

[error log.txt](https://github.com/gramineproject/gramine/files/13300221/error.log.txt)

Thank you.
taserghar

**Reply by kailun-qin** [reply]
@taserghar From the latest error log, I think there was an unexpected installation of nodejs back to v12.22.9 (an unsupported version by sgx-dcap-pccs) when you tried to install `sgx-dcap-pccs` with `apt`, which did not notice that you've already installed nodejs via `snap`.

IMO, first you need a version **above** v16.20.0, say v16.20.2, because the current version check in PCCS install script does not allow the equal version (see [this logic](https://github.com/intel/SGXDataCenterAttestationPrimitives/blob/39989a42bbbb0c968153a47254b6de79a27eb603/QuoteGeneration/pccs/install.sh#L12-L38)) -- this is a bug and I've synced w/ DCAP team to fix it in private, so you may expect a fix later. Additionally, you should avoid installing nodejs via manual copying, but instead give a try via using PPA -- `curl -fsSL https://deb.nodesource.com/setup_16.x | sudo -E bash -` (the way used in the [DCAP install guide](https://www.intel.com/content/www/us/en/developer/articles/guide/intel-software-guard-extensions-data-center-attestation-primitives-quick-install-guide.html) which'll turn to `apt install nodejs` eventually thus should avoid the issue w/ `snap` (sorry I'm not familiar w/ `snap` -- so maybe there is a simple way but IDK).

**Reply by dimakuv** [reply]
Thanks @kailun-qin for the great feedback and finding the bug in the PCCS install script!

**Reply by taserghar** [reply]
Hi @kailun-qin  @dimakuv Thanks for the replies. I tried to install pccs with apt but it seem doesn't work well for me. So I installed it as per the instruction of [DCAP PCCS from git](https://github.com/intel/SGXDataCenterAttestationPrimitives/tree/master/QuoteGeneration/pccs)

But for the configuration, I used same as the same link you provided:
> https://www.intel.com/content/www/us/en/developer/articles/guide/intel-software-guard-extensions-data-center-attestation-primitives-quick-install-guide.html

Finally when I ran the pccs_server the output is like below:
>root@SGX-server:/home/dhar/downloads/SGXDataCenterAttestationPrimitives/QuoteGeneration/pccs# node pccs_server.js 
2023-11-14 14:56:38.299 [info]: HTTPS Server is running on: https://localhost:8081

Also, I installed the QPL as mentioned in the earlier comments. 

@mrinal054  Could you please verify if the attestation is functioning properly now?

**Reply by mrinal054** [reply]
Now, I am getting the following error:

```
(host_main.c:562:initialize_enclave) debug: Added all pages to SGX enclave
(host_framework.c:516:init_enclave) debug: Enclave initializing:
(host_framework.c:517:init_enclave) debug:     enclave id:   0x00000000fffff000
(host_framework.c:518:init_enclave) debug:     mr_enclave:   a220954d95afb78c194e1cee7112462a3d9c6d305f899d23ad08d36748b71ae7
(host_main.c:1039:load_enclave) debug: Using SGX attestation type "dcap"
(host_platform.c:195:init_quoting_enclave_targetinfo) error: AESM service returned error 46; this may indicate that infrastructure for the DCAP attestation requested by Gramine is missing on this machine
(host_main.c:1269:main) error: load_enclave() failed with error: Operation not permitted (EPERM)
```

I am also attaching the detailed log herewith.

[client.txt](https://github.com/gramineproject/gramine/files/13386645/client.txt)

Also, before running the client-side, I run the PCCS server in a separate terminal. Here is the log below:
```
(sgx) dhar@SGX-server:~$ sudo su
[sudo] password for dhar: 
root@SGX-server:/home/dhar# cd /home/dhar/downloads/SGXDataCenterAttestationPrimitives/QuoteGeneration/pccs
root@SGX-server:/home/dhar/downloads/SGXDataCenterAttestationPrimitives/QuoteGeneration/pccs# node pccs_server.js
2023-11-14 08:20:07.202 [info]: HTTPS Server is running on: https://localhost:8081
```


Please note that I can still execute the PyTorch code with Gramine in the SGX enclave without remote attestation.

**Reply by dimakuv** [reply]
There was already a discussion on the error `46`  -- please see https://github.com/gramineproject/gramine/issues/516

**Reply by mrinal054** [reply]
We re-installed and updated drivers. And, followed the suggestion as mentioned in #516. However, still getting some new errors.

Here is my dcap-server side:
```
(sgx) dhar@SGX-server:~/Gramine_project/examples/pytorch-confidential/secret_prov_pf$ ./server_dcap ../files/wrap_key & RA_TLS_ALLOW_OUTDATED_TCB_INSECURE=1 
[1] 126129
(sgx) dhar@SGX-server:~/Gramine_project/examples/pytorch-confidential/secret_prov_pf$ --- Reading the master key for encrypted files from '../files/wrap_key' ---
--- Starting the Secret Provisioning server on port 4433 ---
ra_tls_verify_callback: Quote: verification failed with error OUT_OF_DATE_CONFIG_NEEDED
client_connection: Secret Provisioning failed during mbedtls_ssl_handshake with error -12288
client_connection: ra_tls_verify_callback_results:
    attestation_scheme=2, err_loc=3, 
client_connection:     dcap.func_verify_quote_result=0x0, dcap.quote_verification_result=0xa003
```

Here is the client side:
```
secret_provision_constructor: Secret provisioning failed, terminating the whole process
(libos_parser.c:1628:buf_write_all) [P1:T1:python3.10] trace: ---- return from write(...) = 0x58
(libos_exit.c:212:libos_syscall_exit_group) [P1:T1:python3.10] debug: ---- exit_group (returning 1)
(libos_fs_lock.c:785:file_lock_clear_pid) [P1:T1:python3.10] debug: clearing file (POSIX) locks for pid 1
(libos_sync_client.c:331:shutdown_sync_client) [P1:T1:python3.10] debug: sync client shutdown: closing handles
(libos_sync_client.c:346:shutdown_sync_client) [P1:T1:python3.10] debug: sync client shutdown: waiting for confirmation
(libos_sync_client.c:359:shutdown_sync_client) [P1:T1:python3.10] debug: sync client shutdown: finished
(libos_ipc_worker.c:285:ipc_worker_main) [P1:libos] debug: IPC worker: exiting worker thread
(libos_exit.c:58:libos_clean_and_exit) [P1:T1:python3.10] debug: process 1 exited with status 1
(pal_process.c:248:_PalProcessExit) debug: PalProcessExit: Returning exit code 1
```
[client.txt](https://github.com/gramineproject/gramine/files/13519317/client.txt)


I have also attached the full log to this message.

**Reply by kailun-qin** [reply]
> ra_tls_verify_callback: Quote: verification failed with error OUT_OF_DATE_CONFIG_NEEDED
> dcap.quote_verification_result=0xa003

This indicated that the TCB level of your platform is out of date and additional configuration of the SGX platform at its current patching level may be needed (pls see https://github.com/gramineproject/gramine/blob/24c8e9bfb5a8d931dc1fbd62e50b1a77eeddea78/tools/sgx/ra-tls/ra_tls_verify_dcap.c#L50-L53). To return an OK status, you may firstly want to double check the TCB level and the config of your platform.

But if you do want to allow the not-OK status, you need to explicitly set the environment variables (`RA_TLS_ALLOW_OUTDATED_TCB_INSECURE` and `RA_TLS_ALLOW_HW_CONFIG_NEEDED` for this error), pls see: https://github.com/gramineproject/gramine/blob/24c8e9bfb5a8d931dc1fbd62e50b1a77eeddea78/Documentation/attestation.rst?plain=1#L331-L347.

**Reply by keke-chen** [reply]
Hello, 
  We tried the example under gramine/CI-Examples/ra-tls-secret-prov/secret_prov_pf/
following the instructions about "Secret Provisioning flows, ECDSA-based (DCAP) attestation:"

```
make app dcap RA_TYPE=dcap
cd secret_prov_pf
./server_dcap wrap_key &
gramine-sgx ./client
```

Everything worked fine and I could see the "helloworld" message.
Then, I ran "kill %%" to terminate the server and copied the files to the pytorchconfidential directory, following the steps in the tutorial to encrypt data and compile the SGX version. 

```
...
mkdir secret_prov_pf
cp gramine/CI-Examples/ra-tls-secret-prov/secret_prov_pf/server_dcap secret_prov_pf/
cp -R gramine/CI-Examples/ra-tls-secret-prov/ssl ./
cd secret_prov_pf
./server_dcap ../files/wrap_key &
gramine-sgx ./pytorch pytorchexample.py 2>err.txt
```

It returns the following error messages:
```
secret_provision_start: Secret Provisioning failed during mbedtls_ssl_handshake with error -9984
secret_provision_constructor: Secret provisioning failed, terminating the whole process
client_connection: Secret Provisioning failed during mbedtls_ssl_handshake with error -80
client_connection: ra_tls_verify_callback_results:
    attestation_scheme=0, err_loc=0, 
client_connection:     unknown attestation scheme!
```

The detailed debug information is in the following file
[err.txt](https://github.com/gramineproject/gramine/files/13655586/err.txt)

Dear [kailun-qin](https://github.com/kailun-qin), could you please help identify the problems? Thanks!

**Reply by kailun-qin** [reply]
> Then, I ran "kill %%" to terminate the server and copied the files to the pytorchconfidential directory, following the steps in the tutorial to encrypt data and compile the SGX version.

@keke-chen Before building and running the example, have you also followed the [manifest update steps](https://github.com/gramineproject/gramine/blob/master/Documentation/tutorials/pytorch/index.rst#preparing-manifest-file) required for the remote secret provisioning?

**Reply by keke-chen** [reply]
Yes. 

```
[loader]
entrypoint = "file:/usr/local/lib/x86_64-linux-gnu/gramine/libsysdb.so"
log_level = "all"
insecure__use_cmdline_argv = true

[loader.env]
LD_LIBRARY_PATH = "/lib:/usr/lib:/lib/x86_64-linux-gnu:/usr//lib/x86_64-linux-gnu"
HOME = "/home/keke"
OMP_NUM_THREADS = "8"
LD_PRELOAD = "libsecret_prov_attest.so"
SECRET_PROVISION_CONSTRUCTOR = "1"
SECRET_PROVISION_SET_KEY = "default"
SECRET_PROVISION_CA_CHAIN_PATH = "./ssl/ca.crt"
SECRET_PROVISION_SERVERS = "127.0.0.1:4433"

[libos]
entrypoint = "/usr/bin/python3.10"

[[fs.mounts]]
path = "/usr/bin/python3.10"
uri = "file:/usr/bin/python3.10"

[[fs.mounts]]
path = "/lib"
uri = "file:/usr/local/lib/x86_64-linux-gnu/gramine/runtime/glibc"

[[fs.mounts]]
path = "/usr/lib"
uri = "file:/usr/lib"

[[fs.mounts]]
path = "/lib/x86_64-linux-gnu"
uri = "file:/lib/x86_64-linux-gnu"

[[fs.mounts]]
path = "/usr//lib/x86_64-linux-gnu"
uri = "file:/usr//lib/x86_64-linux-gnu"

[[fs.mounts]]
path = "/usr/lib/python3.10"
uri = "file:/usr/lib/python3.10"

[[fs.mounts]]
path = "/usr/lib/python3.10/lib-dynload"
uri = "file:/usr/lib/python3.10/lib-dynload"

[[fs.mounts]]
path = "/usr/local/lib/python3.10/dist-packages"
uri = "file:/usr/local/lib/python3.10/dist-packages"

[[fs.mounts]]
path = "/usr/lib/python3/dist-packages"
uri = "file:/usr/lib/python3/dist-packages"

[[fs.mounts]]
type = "tmpfs"
path = "/tmp"

[[fs.mounts]]
path = "/classes.txt"
uri = "file:classes.txt"
type = "encrypted"

[[fs.mounts]]
path = "/input.jpg"
uri = "file:input.jpg"
type = "encrypted"

[[fs.mounts]]
path = "/alexnet-pretrained.pt"
uri = "file:alexnet-pretrained.pt"
type = "encrypted"

[[fs.mounts]]
path = "/result.txt"
uri = "file:result.txt"
type = "encrypted"

[sgx]
enclave_size = "4G"
max_threads = 32
edmm_enable = false
trusted_files = [
    { uri = "file:/usr/bin/python3.10" },
    { uri = "file:/usr/local/lib/x86_64-linux-gnu/gramine/libsysdb.so" },
    { uri = "file:/usr/local/lib/x86_64-linux-gnu/gramine/runtime/glibc/" },
    { uri = "file:/usr/lib/" },
    { uri = "file:/lib/x86_64-linux-gnu/" },
    { uri = "file:/usr//lib/x86_64-linux-gnu/" },
    { uri = "file:/usr/lib/python3.10/" },
    { uri = "file:/usr/lib/python3.10/lib-dynload/" },
    { uri = "file:/usr/local/lib/python3.10/dist-packages/" },
    { uri = "file:/usr/lib/python3/dist-packages/" },
    { uri = "file:pytorchexample.py" },
    { uri = "file:ssl/ca.crt" },
]
remote_attestation = "dcap"
isvprodid = 0
isvsvn = 0
debug = false
enable_stats = false
use_exinfo = false

[sgx.cpu_features]
avx = "unspecified"
avx512 = "unspecified"
amx = "unspecified"
mpx = "disabled"
pkru = "disabled"

[sys]
enable_extra_runtime_domain_names_conf = true
```

**Reply by kailun-qin** [reply]
Thanks!

> secret_provision_start: Secret Provisioning failed during mbedtls_ssl_handshake with error -9984

`-9984` (`-0x2700`) indicated that the certificate verification failed, e.g. CRL, CA or signature check failed (pls see: https://github.com/Mbed-TLS/mbedtls/blob/1fdc884ed831927600838a3656b6d74a0417d52b/include/mbedtls/x509.h#L83).

However, there are several reasons for certificate verification to fail (e.g., incorrect trusted root CA certificate, the server hostname is different from the certificate CN in SN\SAN...). Since you have both the client and server running on the same localhost and you are able to run `ra-tls-secret-prov` example correctly and you directly copied the related files, I don't see anything suspicious obviously and can't tell w/o further debug info. Any other context I may have missed?

**Reply by dimakuv** [reply]
@keke-chen This is suspicious in your output:
```
client_connection: ra_tls_verify_callback_results:
    attestation_scheme=0, err_loc=0, 
client_connection:     unknown attestation scheme!
```

Somehow the secret provisioning server (`./server_dcap`) doesn't understand the your Gramine-SGX PyTorch client sent a *DCAP* attestation quote. Are you sure you have a "classic" SGX attestation infrastructure? Could your SGX infrastructure be based not on Intel DCAP, but on something else (like Microsoft Azure Attestation)?

**Reply by kailun-qin** [reply]
@dimakuv I think this can happen when mbedtls handshake failed early and not even reached the RA-TLS verify callback (i.e., irrelevant to the attestation scheme)?

**Reply by keke-chen** [reply]
Here is the output and log of successfully running gramine/CI-Examples/ra-tls-secret-prov/secret_prov_pf
It's on the same server -- I copied server_dcap and the ssl/ directory to pytorchconfidential as instructed

=====
gramine-sgx ./client 2>err1.txt                                                   
Allowing quote status CONFIG_AND_SW_HARDENING_NEEDED
Received the following measurements from the client:
  - MRENCLAVE:   21a2e9dbf1491f893aae94b17ca0de7ea9592603da8eaf1c250018b30bfb6d1a
  - MRSIGNER:    70199cf5824df64a4dd601416ee149d813c138afdb8c3eb8945be46b1371ec2d
  - ISV_PROD_ID: 0
  - ISV_SVN:     0
[ WARNING: In reality, you would want to compare against expected values! ]
--- [main] Re-starting myself using execvp() ---
--- [child] Read from protected file: 'helloworld' ---
--- [parent] Read from protected file: 'helloworld' ---

[err1.txt](https://github.com/gramineproject/gramine/files/13661912/err1.txt)

**Reply by dimakuv** [reply]
Hm, this information about a successful run of `secret_prof_pf` example doesn't help much. It just shows that your DCAP infrastructure seems to be installed correctly and works fine.

Why you see problems with the PyTorch example, I still can't understand.
